### PR TITLE
new dispatch middleware

### PR DESCRIPTION
**Pull request by CordlessWool**
## Proposed Changes
  -
  -
  -

## ToDos
- [ ] this PR targets `base:development`
- [ ] referenced corresponding issue (if any)
- [ ] ...

Fixes #

### CODE REVIEW

File: src/actions/lesson.jsx [hunk 1]
```diff
@@ -3,8 +3,9 @@ import { useContext } from "react"
 import LessonContext from "~/Contexts/Lesson"
 import uuid from "uuid/v4"
 
-export const createSection = (dispatch) => async (lessonId, position) => {
+export const createSection = (position) => async ({dispatch, state}) => {
 	const tempId = uuid()
```
Comment (CeEv) [inline src/actions/lesson.jsx]: Ist das für die Position? Wenn ja dann würde ich das auch so benennen.

File: src/actions/lesson.jsx [hunk 2]
```diff
@@ -3,8 +3,9 @@ import { useContext } from "react"
 import LessonContext from "~/Contexts/Lesson"
 import uuid from "uuid/v4"
 
-export const createSection = (dispatch) => async (lessonId, position) => {
+export const createSection = (position) => async ({dispatch, state}) => {
```
Comment (CeEv) [inline src/actions/lesson.jsx]: {dispatch: lessonDispatcher, state: lessonState} ?

Comment (CeEv) [reply src/actions/lesson.jsx]: /**
 @params {Object} options {dispatch, state}
**

File: src/Contexts/Lesson.jsx [hunk 3]
```diff
@@ -347,10 +348,10 @@ function reducer(state, { type, payload }) {
 
 const LessonContext = React.createContext()
 
-export function LessonContextProvider({ children }) {
+export function LessonContextProvider({ children, dispatchMiddleware = [] }) {
     const [state, dispatch] = useReducer(reducer, initialState)
 
-    const value = { store: state, dispatch }
+    const value = { store: state, dispatch: createDispatch([thunkMiddleware], dispatch, state) }
```
Comment (CeEv) [inline src/Contexts/Lesson.jsx]: "thunk"Middleware ?

File: src/Contexts/Lesson.jsx [hunk 4]
```diff
@@ -347,10 +348,10 @@ function reducer(state, { type, payload }) {
 
 const LessonContext = React.createContext()
 
-export function LessonContextProvider({ children }) {
+export function LessonContextProvider({ children, dispatchMiddleware = [] }) {
```
Comment (CeEv) [inline src/Contexts/Lesson.jsx]: dispatchMiddleware  wird aktuell nicht verwendet

File: src/Contexts/Lesson.jsx [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import React, { useReducer } from "react"
 import qs from "qs"
 import { mergeDiff } from "~/utils/diff"
+import { createDispatch, thunkMiddleware } from "~/utils/dispatch"
```
Comment (CeEv) [inline src/Contexts/Lesson.jsx]: createDispatch ist keine Middleware oder? Middleware unter utils finde ich auch schwierig.

File: src/actions/lesson.jsx (hunk 1 above)
Comment (CordlessWool) [reply src/actions/lesson.jsx]: Ne ist tatsächlich um sections zu erstellen, kannst dabei auch auch direkt die position mitgeben

File: src/utils/dispatch.test.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+const reducer = (state, {type, payload}) => {
+	switch(type){
+		case 'MIDDLEWARE':
+			return {
+				...state,
+				test: payload
+			}
+		case 'DEFAULT':
```
Comment (CeEv) [inline src/utils/dispatch.test.js]: The case DEFAULT and default case in line 46, maybe rename DEFAULT in any other type name.

File: src/utils/dispatch.test.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+	const rendered = renderer.create(<Provider />)
+	const renderedInstance = rendered.root
+
+	const comp = renderedInstance.findByType(TestComponent)
+	const {state, dispatch} = comp.props
+
+	const switchState = state.switch
+	dispatch(asyncFu('Affenmesserkampf!'))
+	setTimeout(() => {
```
Comment (CeEv) [inline src/utils/dispatch.test.js]: Please add comment, why it must explicit exist a timeout for this test.

Comment (CeEv) [reply src/utils/dispatch.test.js]: Ok, but what you really want is such as await dispatch(asyncFu());
https://github.com/jeffbski/redux-logic/issues/88
Maybe we should wrap dispatch that include async operations with a global util like process / processHandler that enforce that the return value is a Promise that include as resolve the dispatcher and other return values.

Comment (CordlessWool) [reply src/utils/dispatch.test.js]: I tried it with await, did not get it work

**Pull request merged**